galley/pkg/resource/synchronizer.go

```diff
@@ -0,0 +1,239 @@
+func NewSynchronizer(config *rest.Config, resyncPeriod time.Duration, name string, source schema.GroupVersion,
+	destination schema.GroupVersion, kind string, listKind string) (s *Synchronizer, err error) {
+	return newSynchronizer(config, resyncPeriod, name, source, destination, kind, listKind, nil)
+}
+
+// NewSynchronizer returns a new instance of a Synchronizer that synchronizes the custom resource contents
+// from the accessor APIGroup/Version to the destination APIGroup/Version. The kind and listKind is used
+// when hydrating objects as unstructured.Unstructured.
+func newSynchronizer(
+	config *rest.Config, resyncPeriod time.Duration, name string, source schema.GroupVersion,
+	destination schema.GroupVersion, kind string, listKind string, eventHook eventHookFn) (
+	s *Synchronizer, err error) {
+
+	s = &Synchronizer{
+		eventHook: eventHook,
+	}
+
+	if s.source, err = newAccessor(config, resyncPeriod, name,
+		source, kind, listKind, s.handleInputEvent); err != nil {
+
+		s = nil
+		return
+	}
+
+	if s.destination, err = newAccessor(config, resyncPeriod, name,
+		destination, kind, listKind, s.handleOutputEvent); err != nil {
+
+		s = nil
+		return
+	}
+
+	return
+}
+
+// Start the resource synchronizer
+func (s *Synchronizer) Start() {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	if s.running {
+		log.Errorf("Synchronizer has already started: %s", s.source.name)
+		return
+	}
+
+	log.Infof("Starting resource synchronization: %s(%s) => %s(%s)",
+		s.source.name, s.source.gv, s.destination.name, s.destination.gv)
+
+	s.queue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), s.source.name)
+	s.source.start()
+	s.destination.start()
+
+	s.waitForProcess.Add(1)
+	go s.process()
+
+	s.running = true
+}
+
+// Stop the resource synchronizer.
+func (s *Synchronizer) Stop() {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	if !s.running {
+		log.Error("Already stopped")
+		return
+	}
+
+	s.source.stop()
+	s.destination.stop()
+
+	s.queue.ShutDown()
+	s.waitForProcess.Wait()
+
+	s.running = false
+}
+
+// background operation loop
+func (s *Synchronizer) process() {
+	for {
+		item, shutdown := s.queue.Get()
+		if shutdown {
+			break
+		}
+
+		if info, ok := item.(*change.Info); !ok {
+			log.Errorf("Got a non-change item from the queue: %v", item)
+		} else if s.processChange(info) {
+			s.queue.Forget(item)
+			log.Debugf("item processing complete successfully: %v", item)
+		}
+
+		if s.eventHook != nil {
+			s.eventHook(item)
+		}
+		s.queue.Done(item)
+	}
+
+	s.waitForProcess.Done()
+}
+
+func (s *Synchronizer) processChange(info *change.Info) bool {
+	var sourceRes *unstructured.Unstructured
+	var destinationRes *unstructured.Unstructured
+	var err error
+
+	if sourceRes, err = getResource(s.source, info.Name); err != nil {
+		return false
+	}
+	if destinationRes, err = getResource(s.destination, info.Name); err != nil {
+		return false
+	}
```

Comment:
Would the following be more concise and avoid the forward declarations? 

```golang
sourceRes, err := getResource(s.source, info.Name)
if err != nil {
  return false
}
destinationRes, err = getResource(s.destination, info.Name)
if  err != nil {
	return false
}
```

Reply:
I don't think it saves much and would end up redeclaring the err. I don't care either way much, but in this instance this seems a bit more consistent.

Reply:
I don't see re-declaring `err` as a big deal, but linters may disagree =/